src/it/process-jar/pom.xml

```diff
@@ -0,0 +1,115 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.jenkins-ci.plugins</groupId>
+    <artifactId>plugin</artifactId>
+    <version>1.424</version>
+  </parent>
+
+  <groupId>org.jenkins-ci.tools.hpi.its</groupId>
+  <artifactId>process-jar</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <packaging>hpi</packaging>
+
+  <name>MyNewPlugin</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+      <version>2.6.3</version>
+      <optional>true</optional>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <groupId>@project.groupId@</groupId>
+          <artifactId>@project.artifactId@</artifactId>
+          <version>@project.version@</version>
+        </plugin>
+        <plugin>
+          <groupId>org.kohsuke</groupId>
+          <artifactId>access-modifier-checker</artifactId>
+          <version>1.0</version>
+        </plugin>
+      </plugins>
+    </pluginManagement>
+    <plugins>
+      <plugin>
+        <groupId>@project.groupId@</groupId>
+        <artifactId>@project.artifactId@</artifactId>
+        <version>@project.version@</version>
+        <executions>
+          <execution>
+            <phase>prepare-package</phase>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+            <!-- TODO the shade plugin doesn't like shading the non-main attached artifacts yet -->
+            <!--<configuration>-->
+              <!--<jarClassifier>unshaded</jarClassifier>-->
+            <!--</configuration>-->
+          </execution>
+        </executions>
+        <configuration>
+          <jarClassifier>shaded</jarClassifier>
+        </configuration>
+      </plugin>
+      <!-- TODO the shade plugin doesn't like shading the non-main attached artifacts yet -->
```

Comment:
@KostyaSha w.r.t. 3

What is actually happening is that when you add `jar:jar` to the lifecycle (which you currently need in order to let `shade:shade` work) then the resulting `jar` file is attached as the main artifact... and then attempted to be replaced as the main artifact by `shade:shade`... then along comes `hpi:hpi` which produces its own `.jar` and tries to set the main artifact to the `.hpi` file...

Depending on who sets the main artifact first and which version of Maven you are running you can thus end up with any of three different artifacts as the main artifact, plus the `.jar` that `hpi:hpi` produced being added as a side attached artifact...

There is a secondary issue with producing the shaded jar file in the same reactor as consuming it when running Maven 3.3.x and where you create a dependency reduced pom... the work-around for that issue is to make the pom such that it is "effectively" dependency reduced already by exploiting `<optional>true</optional>` for all the dependencies that you want to remove